livekit-plugins/livekit-plugins-assemblyai/livekit/plugins/assemblyai/stt.py

```diff
@@ -0,0 +1,395 @@
+                    data = json.loads(msg.data)
+                    self._process_stream_event(data)
+                except Exception:
+                    logger.error("failed to process AssemblyAI message")
+
+        await asyncio.gather(send_task(), recv_task())
+
+    def _end_speech(self) -> None:
+        if not self._speaking:
+            logger.info(
+                "trying to commit final events without being in the speaking state",
+            )
+            return
+
+        if len(self._final_events) == 0:
+            return
+
+        self._speaking = False
+
+        # combine all final transcripts since the start of the speech
+        sentence = ""
+        confidence = 0.0
+        for f in self._final_events:
+            alt = f.alternatives[0]
+            sentence += f"{alt.text.strip()} "
+            confidence += alt.confidence
+
+        sentence = sentence.rstrip()
+        confidence /= len(self._final_events)  # avg. of confidence
+
+        end_event = stt.SpeechEvent(
+            type=stt.SpeechEventType.END_OF_SPEECH,
+            alternatives=[
+                stt.SpeechData(
+                    language=ENGLISH,
+                    start_time=self._final_events[0].alternatives[0].start_time,
+                    end_time=self._final_events[-1].alternatives[0].end_time,
+                    confidence=confidence,
+                    text=sentence,
+                ),
+            ],
+        )
+        self._event_queue.put_nowait(end_event)
+        self._final_events = []
+
+    def _process_stream_event(self, data: dict) -> None:
+        # see this page:
+        # https://www.assemblyai.com/docs/api-reference/streaming/realtime
+        # for more information about the different types of events
+        if data["message_type"] == "SessionBegins":
+            self._speaking = True
+            start_event = stt.SpeechEvent(type=stt.SpeechEventType.START_OF_SPEECH)
+            self._event_queue.put_nowait(start_event)
+
+        elif data["message_type"] == "PartialTranscript":
+            self._speaking = True
+            alts = live_transcription_to_speech_data(ENGLISH, data)
+            if len(alts) > 0 and alts[0].text:
+                interim_event = stt.SpeechEvent(
+                    type=stt.SpeechEventType.INTERIM_TRANSCRIPT,
+                    alternatives=alts,
+                )
+                self._event_queue.put_nowait(interim_event)
+
+        elif data["message_type"] == "FinalTranscript":
+            alts = live_transcription_to_speech_data(ENGLISH, data)
+            if len(alts) > 0 and alts[0].text:
+                final_event = stt.SpeechEvent(
+                    type=stt.SpeechEventType.FINAL_TRANSCRIPT,
+                    alternatives=alts,
+                )
+                self._final_events.append(final_event)
+                self._event_queue.put_nowait(final_event)
+                self._end_speech()
```

Comment:
I don't think this is correct. We don't want to end the speech for every final transcript. a final transcript can be just a final sentence but not the end of the speech